### ISSUE TITLE
holes in metadata

### ISSUE THREAD

**Issue by borkd** [borgbackup/borg#4867]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

BUG

#### Your borg version (borg -V).
1.1.9 and 1.1.10

#### Operating system (distribution) and version.

Debian 9.11

#### Hardware / network configuration, and filesystems used.

tmpfs, ZFS, NFS

#### How much data is handled by borg?

Gigabytes to Terabytes

## Describe the problem you're observing.

I expect that all metadata handled by borg can be saved during create, and later presented in a reproducible manner regardless on how the archive is later consumed. For example `borg list`, `borg mount`, and `borg extract` should produce identical filesystem trees.

This is not the case when backed up content is "cherry-picked" from a within a directory tree.

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

`# borg [common options] create [options] ARCHIVE /a /b/c /d/f/{g,h}` will save correct metdata 
 for `/a` `/b/c`, `/d/f/g`, `/d/f/h`, and their descendants, but will leave holes in the metadata for `/b`, `/d`, and `/d/f`

Unlike `tar`, `borg list` does not produce contiguous listing, quietly ignoring any missing parent directories that make up the holes. 

Problem becomes obvious once one proceeds to mount the archive, or restore the data and compare the resulting filesystem tree with the source.

Mount and extract quietly make up missing parent directory entries with operator UID as owner, effective umask for permissions, and current (and ever changing) timestamp.

For consistency and reproducibility sake, is it possible to store metadata for parent directories as well?

#### Include any warning/errors/backtraces from the system logs

No errors reported

**Comment by ThomasWaldmann** [comment]
IIRC there is / was already an issue about this.

borg backs up exactly what you say.

If you say "backup b/c", it will start the recursion at `b/c` (which means the `c` item, to be precise) and back up that and everything below that (if it is a directory). The `b` directory item is not part of your backup then.

You could also start the recursion at `b`, then `b` would be part of your backup (and you could exclude stuff you do not need).

Similarly, if you `extract b/c` you tell borg to extract the item with exactly that path (which is the `c` item, to be precise). If the `b` path component does not already exist, borg auto-creates it (this is different from extracting a directory item), so it can extract the `b/c` item to correct relative path.

If you wanted to extract (not auto-create) the `b` item, you need to `extract b`.

**Comment by borkd** [comment]
My most frequent (ab)use case composes `borg create` command from a list of artifact regexes I may be interested in later, and saving the results to the closest available "thin" repo under a deterministic, but unique archive name. Sometimes I need to "sync" the archives between thin repositories participating in the given experiment to post-process the data. Eventually the scattered archives are sent to a central repository.

In the post-processing step, reproducibility tests are executed against the borgfs mountpoints, and do pairwise comparisons of archives, looking for differences in contents and metadata, starting at the mountpoint. Possibility of encountering made up and ever changing metadata due to holes makes scripting around this rather fragile.

TL;DR
Awaiting for native tar import feature (making borg send|receive a thing), I will experiment with more elaborate includes/excludes to try to capture the metadata of the parent directories

**Comment by ThomasWaldmann** [comment]
At the moment of extracting parent/child, sequential processing has already passed by parent, so it does not have the correct metadata at hand.

Guess the only solution possible would be to keep a cache of directory metadata in memory, so it can be looked up at ad-hoc creation time.

**Comment by ThomasWaldmann** [comment]
duplicate of #1751.
